arbor/network_impl.cpp

```diff
@@ -0,0 +1,388 @@
+
+                    // check all synapses of cell for potential destination
+
+                    for (const auto& [_, placed_synapses]: cell.synapses()) {
+                        for (const auto& p_syn: placed_synapses) {
+                            const auto& label = lid_to_label(cell.synapse_ranges(), p_syn.lid);
+
+                            if (selection.select_destination(cell_kind::cable, gid, label)) {
+                                const mpoint point = location_resolver.at(p_syn.loc);
+                                std::lock_guard<std::mutex> guard(dest_sites_mutex);
+                                dest_sites.insert(
+                                    {gid, p_syn.lid, cell_kind::cable, label, p_syn.loc, point});
+                            }
+                        }
+                    }
+
+                    // check all detectors of cell for potential source
+                    for (const auto& p_det: cell.detectors()) {
+                        const auto& label = lid_to_label(cell.detector_ranges(), p_det.lid);
+                        if (selection.select_source(cell_kind::cable, gid, label)) {
+                            const mpoint point = location_resolver.at(p_det.loc);
+                            std::lock_guard<std::mutex> guard(src_sites_mutex);
+                            src_sites.insert(
+                                {gid, p_det.lid, cell_kind::cable, label, p_det.loc, point});
+                        }
+                    }
+                });
+        }
+        else {
+            // Assuming all other cell types do not have a morphology. We can use label
+            // resolution through factory and set local position to 0.
+            auto factory = cell_kind_implementation(kind, backend_kind::multicore, *ctx, 0);
+
+            // We only need the label ranges
+            cell_label_range sources, destinations;
+            std::ignore = factory(gids, rec, sources, destinations);
+
+            std::size_t source_label_offset = 0;
+            std::size_t destination_label_offset = 0;
+            for (std::size_t i = 0; i < gids.size(); ++i) {
+                const auto gid = gids[i];
+                const auto iso = rec.get_cell_isometry(gid);
+                const auto point = iso.apply(mpoint{0.0, 0.0, 0.0, 0.0});
+                const auto num_source_labels = sources.sizes().at(i);
+                const auto num_destination_labels = destinations.sizes().at(i);
+
+                // Iterate over each source label for current gid
+                for (std::size_t j = source_label_offset;
+                     j < source_label_offset + num_source_labels;
+                     ++j) {
+                    const auto& label = sources.labels().at(j);
+                    const auto& range = sources.ranges().at(j);
+                    for (auto lid = range.begin; lid < range.end; ++lid) {
+                        if (selection.select_source(kind, gid, label)) {
```

Comment:
Instead of taking a lock here, we could produce and array if outputs and merge them (aka map/reduce) as done in fvm_layout?!

Reply:
I did some initial testing and instantiating cells out-shadowed the cost of locks. But it turns out, that for the later sampling stage, the overhead of locks is quite high. So I've implemented your suggestion now.